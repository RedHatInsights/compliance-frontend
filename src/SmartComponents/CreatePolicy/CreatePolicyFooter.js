import { Button, ButtonVariant } from '@patternfly/react-core';
import {
  WizardContextConsumer,
  WizardFooter,
} from '@patternfly/react-core/deprecated';
import React from 'react';

const CreatePolicyFooter = () => (
  <WizardFooter>
    <WizardContextConsumer>
      {({ onNext, onBack, onClose, activeStep }) => {
        const stepNameEdited = activeStep.name
          .toLowerCase()
          .replaceAll(' ', '-');

        return (
          <>
            <Button
              id={`${stepNameEdited}-next-button`}
              variant={ButtonVariant.primary}
              type="submit"
              onClick={onNext}
              isDisabled={
                !(activeStep && activeStep.enableNext !== undefined
                  ? activeStep.enableNext
                  : true)
              }
            >
              {activeStep.nextButtonText || 'Next'}
            </Button>
            {!activeStep.hideBackButton && (
              <Button
                id={`${stepNameEdited}-back-button`}
                variant={ButtonVariant.secondary}
                onClick={onBack}
                isDisabled={activeStep.id === 1}
              >
                {activeStep.backButtonText || 'Back'}
              </Button>
            )}
            {!activeStep.hideCancelButton && (
              <div className="pf-v5-c-wizard__footer-cancel">
                <Button
                  id={`${stepNameEdited}-cancel-button`}
                  variant={ButtonVariant.link}
                  onClick={onClose}
                >
                  {activeStep.cancelButtonText || 'Cancel'}
                </Button>
              </div>
            )}
          </>
        );
      }}
    </WizardContextConsumer>
  </WizardFooter>
);

export default CreatePolicyFooter;
