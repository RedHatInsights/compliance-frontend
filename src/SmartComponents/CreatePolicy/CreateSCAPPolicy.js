import React from 'react';
import propTypes from 'prop-types';
import {
  Form,
  FormGroup,
  Text,
  TextContent,
  TextVariants,
  Tile,
} from '@patternfly/react-core';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import Spinner from '@redhat-cloud-services/frontend-components/Spinner';
import {
  propTypes as reduxFormPropTypes,
  formValueSelector,
  reduxForm,
} from 'redux-form';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { StateViewPart, StateViewWithError } from 'PresentationalComponents';
import PolicyTypesTable from './Components/PolicyTypeTable';
import PolicyTypeTooltip from './Components/PolicyTypeTooltip';

const SUPPORTED_PROFILES = gql`
  query supportedProfilesByOSMajor {
    osMajorVersions {
      edges {
        node {
          osMajorVersion
          profiles {
            id
            name
            refId
            description
            supportedOsVersions
            benchmark {
              id
              refId
            }
          }
        }
      }
    }
    profiles(search: "external = false and canonical = false") {
      edges {
        node {
          id
          refId
          benchmark {
            refId
          }
        }
      }
    }
  }
`;

export const CreateSCAPPolicy = ({
  change,
  selectedProfile,
  selectedOsMajorVersion,
}) => {
  const { data, error, loading } = useQuery(SUPPORTED_PROFILES, {
    fetchPolicy: 'no-cache',
  });
  const isInUse = (profileRefId, benchmarkRedId) =>
    !!data?.profiles?.edges
      .map(({ node }) => node)
      .find(
        (profile) =>
          profile.refId === profileRefId &&
          benchmarkRedId === profile.benchmark.refId
      );
  const osMajorVersions = data?.osMajorVersions?.edges.map(({ node }) => node);
  const selectedOsMajorVersionObject = osMajorVersions?.find(
    ({ osMajorVersion }) => osMajorVersion === selectedOsMajorVersion
  );
  const profilesToSelect = selectedOsMajorVersionObject?.profiles.map(
    (profile) => ({
      ...profile,
      disabled: isInUse(profile.refId, profile.benchmark.refId),
    })
  );

  return (
    <StateViewWithError stateValues={{ error, data, loading }}>
      <StateViewPart stateKey="loading">
        <Spinner />
      </StateViewPart>
      <StateViewPart stateKey="data">
        <TextContent>
          <Text component={TextVariants.h1} className="pf-u-mb-md">
            Create SCAP policy
          </Text>
          <Text className="pf-u-mb-md">
            Select the operating system and policy type for this policy.
          </Text>
        </TextContent>
        <Form>
          <FormGroup label="Operating system" isRequired fieldId="benchmark">
            {osMajorVersions &&
              osMajorVersions.map(({ osMajorVersion }) => (
                <Tile
                  key={`rhel${osMajorVersion}-select`}
                  className="pf-u-mr-md"
                  title={`RHEL ${osMajorVersion}`}
                  onClick={() => {
                    change('osMajorVersion', osMajorVersion);
                  }}
                  isSelected={selectedOsMajorVersion === osMajorVersion}
                  isStacked
                />
              ))}
          </FormGroup>

          {selectedOsMajorVersion && (
            <FormGroup
              isRequired
              labelIcon={<PolicyTypeTooltip />}
              label="Policy type"
              fieldId="policy-type"
            >
              <PolicyTypesTable
                aria-label="PolicyTypeTable"
                profiles={profilesToSelect}
                onChange={(policy) => {
                  change('profile', policy);
                  change('benchmark', policy.benchmark.id);
                  change('selectedRuleRefIds', undefined);
                  change('systems', []);
                }}
                selectedProfile={selectedProfile}
              />
            </FormGroup>
          )}
        </Form>
      </StateViewPart>
    </StateViewWithError>
  );
};

CreateSCAPPolicy.propTypes = {
  change: reduxFormPropTypes.change,
  selectedProfile: propTypes.object,
  selectedOsMajorVersion: propTypes.string,
};

const selector = formValueSelector('policyForm');

export default compose(
  connect((state) => ({
    selectedProfile: selector(state, 'profile'),
    selectedOsMajorVersion: selector(state, 'osMajorVersion'),
  })),
  reduxForm({
    form: 'policyForm',
    destroyOnUnmount: false,
    forceUnregisterOnUnmount: true,
  })
)(CreateSCAPPolicy);
