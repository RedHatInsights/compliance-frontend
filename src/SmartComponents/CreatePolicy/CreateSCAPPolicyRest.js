import React from 'react';
import propTypes from 'prop-types';
import { propTypes as reduxFormPropTypes } from 'redux-form';
import CreateSCAPPolicyBase from './CreateSCAPPolicyBase';
import useQuery, { apiInstance } from '../../Utilities/hooks/useQuery';
import dataSerialiser from '../../Utilities/dataSerialiser';
import useSupportedProfiles from './useSupportedProfiles';

const profilesDataMap = {
  id: 'id',
  title: 'name',
  ref_id: 'refId',
  security_guide_id: 'benchmark.id',
  supportedOsVersions: 'supportedOsVersions',
};

const serialiseOsVersions = (profiles = []) =>
  profiles.map((profile) => ({
    ...profile,
    supportedOsVersions: profile.os_minor_versions.map(
      (minorVersion) => `${profile.os_major_version}.${minorVersion}`
    ),
  }));

const CreateSCAPPolicyRest = ({
  selectedOsMajorVersion,
  selectedProfile,
  change,
}) => {
  const {
    data: availableOsMajorVersions,
    error: availableOsMajorVersionsError,
    loading: availableOsMajorVersionsLoading,
  } = useQuery(apiInstance.securityGuidesOS);
  const [availableProfiles, availableProfilesError, availableProfilesLoading] =
    useSupportedProfiles(
      selectedOsMajorVersion,
      selectedOsMajorVersion === undefined
    );

  return (
    <CreateSCAPPolicyBase
      {...{
        selectedOsMajorVersion,
        selectedProfile,
        data:
          availableOsMajorVersions === undefined &&
          availableProfiles === undefined
            ? undefined
            : {
                availableOsMajorVersions,
                availableProfiles: dataSerialiser(
                  serialiseOsVersions(availableProfiles),
                  profilesDataMap
                ),
              },
        loading: availableOsMajorVersionsLoading && availableProfilesLoading,
        error: availableOsMajorVersionsError || availableProfilesError,
        change,
      }}
    />
  );
};

CreateSCAPPolicyRest.propTypes = {
  change: reduxFormPropTypes.change,
  selectedProfile: propTypes.object,
  selectedOsMajorVersion: propTypes.string,
};

export default CreateSCAPPolicyRest;
