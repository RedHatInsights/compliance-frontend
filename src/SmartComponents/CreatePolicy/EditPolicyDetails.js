import React, { useEffect } from 'react';
import { compose } from 'redux';
import {
  Field,
  reduxForm,
  formValueSelector,
  propTypes as reduxFormPropTypes,
} from 'redux-form';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import {
  Form,
  FormGroup,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import {
  ReduxFormTextInput,
  ReduxFormTextArea,
} from 'PresentationalComponents/ReduxFormWrappers/ReduxFormWrappers';
import {
  ProfileThresholdField,
  PolicyBusinessObjectiveTooltip,
} from 'PresentationalComponents';

export const EditPolicyDetails = ({ change, policy, refId }) => {
  useEffect(() => {
    if (policy && policy.refId !== refId) {
      change('name', `${policy.name}`);
      change('refId', `${policy.refId}`);
      change('description', `${policy.description}`);
    }
  }, [policy]);

  return (
    <React.Fragment>
      <TextContent>
        <Text component={TextVariants.h1}>Details</Text>
      </TextContent>
      <br />
      <Form id="editpolicydetails">
        <FormGroup label="Policy name" isRequired fieldId="name">
          <Field
            component={ReduxFormTextInput}
            type="text"
            isRequired={true}
            id="name"
            name="name"
            aria-describedby="name"
          />
        </FormGroup>
        <FormGroup label="Reference ID" isRequired fieldId="refId">
          <Field
            type="text"
            component={ReduxFormTextInput}
            isDisabled
            id="refId"
            name="refId"
            aria-describedby="refId"
          />
        </FormGroup>
        <FormGroup label="Description" fieldId="description">
          <Field
            type="text"
            component={ReduxFormTextArea}
            id="description"
            name="description"
            aria-describedby="description"
          />
        </FormGroup>
        <FormGroup
          label="Business objective"
          labelIcon={<PolicyBusinessObjectiveTooltip />}
          fieldId="businessObjective"
        >
          <Field
            type="text"
            component={ReduxFormTextInput}
            id="businessObjective"
            name="businessObjective"
            aria-describedby="businessObjective"
            defaultValue={policy.businessObjective}
          />
        </FormGroup>
        <ProfileThresholdField previousThreshold={100} />
      </Form>
    </React.Fragment>
  );
};

const selector = formValueSelector('policyForm');

EditPolicyDetails.propTypes = {
  policy: propTypes.object,
  refId: propTypes.string,
  change: reduxFormPropTypes.change,
};

const mapStateToProps = (state) => {
  const policy = selector(state, 'profile');
  return {
    policy,
    refId: selector(state, 'refId'),
    initialValues: {
      name: `${policy.name}`,
      refId: `${policy.refId}`,
      description: `${policy.description}`,
      benchmark: selector(state, 'benchmark'),
      osMajorVersion: selector(state, 'osMajorVersion'),
      profile: selector(state, 'profile'),
    },
  };
};

export default compose(
  connect(mapStateToProps),
  reduxForm({
    form: 'policyForm',
    destroyOnUnmount: false,
    forceUnregisterOnUnmount: true,
  })
)(EditPolicyDetails);
