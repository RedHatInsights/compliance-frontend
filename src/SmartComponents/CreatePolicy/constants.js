import gql from 'graphql-tag';

export const BENCHMARKS_QUERY = gql`
  query CP_Benchmarks($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        ruleTree
        valueDefinitions {
          defaultValue
          description
          id
          refId
          title
          valueType
        }
        profiles {
          id
          refId
          osMajorVersion
        }
        version
      }
    }
  }
`;

export const BENCHMARKS_RULES_TREES_QUERY = gql`
  query CP_BenchmarksRuleTree($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        ruleTree
      }
    }
  }
`;

export const BENCHMARKS_VALUE_DEFINITIONS_QUERY = gql`
  query CP_BenchmarksValueDefinitions($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        ruleTree
        valueDefinitions {
          defaultValue
          description
          id
          refId
          title
          valueType
        }
        profiles {
          id
          refId
          osMajorVersion
        }
        version
      }
    }
  }
`;

export const PROFILES_QUERY = gql`
  query CP_Profiles($filter: String!) {
    profiles(search: $filter) {
      edges {
        node {
          id
          refId
          osMinorVersion
          benchmark {
            id
            latestSupportedOsMinorVersions
          }
          rules {
            refId
          }
        }
      }
    }
  }
`;
