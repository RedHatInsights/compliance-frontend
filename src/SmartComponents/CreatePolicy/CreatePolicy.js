import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { formValueSelector } from 'redux-form';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import { Wizard } from '@patternfly/react-core';
import CreateSCAPPolicy from './CreateSCAPPolicy';
import EditPolicyRules from './EditPolicyRules';
import EditPolicySystems from './EditPolicySystems';
import EditPolicyDetails from './EditPolicyDetails';
import ReviewCreatedPolicy from './ReviewCreatedPolicy';
import FinishedCreatePolicy from './FinishedCreatePolicy';
import { validateBenchmarkPage, validateDetailsPage, validateRulesPage } from './validate';

export const CreatePolicy = ({
    benchmark, osMajorVersion, complianceThreshold, name, profile, refId, selectedRuleRefIds
}) => {
    const history = useHistory();
    const [stepIdReached, setStepIdReached] = useState(1);
    const onNext = ({ id }) => {
        setStepIdReached(stepIdReached < id ? id : stepIdReached);
    };

    const onClose = () => {
        history.push('/scappolicies');
    };

    const steps = [
        {
            id: 1,
            name: 'Create SCAP policy',
            component: <CreateSCAPPolicy/>,
            enableNext: validateBenchmarkPage(benchmark, osMajorVersion, profile)
        },
        {
            id: 2,
            name: 'Details',
            component: <EditPolicyDetails/>,
            canJumpTo: stepIdReached >= 2,
            enableNext: validateDetailsPage(name, refId, complianceThreshold)
        },
        {
            id: 3,
            name: 'Systems',
            component: <EditPolicySystems/>,
            canJumpTo: stepIdReached >= 3
        },
        {
            id: 4,
            name: 'Rules',
            component: <EditPolicyRules/>,
            canJumpTo: stepIdReached >= 4,
            enableNext: validateRulesPage(selectedRuleRefIds)
        },
        {
            id: 5,
            name: 'Review',
            component: <ReviewCreatedPolicy/>,
            nextButtonText: 'Finish',
            canJumpTo: stepIdReached >= 5
        },
        {
            id: 6,
            name: 'Finished',
            component: <FinishedCreatePolicy onWizardFinish={ onClose } />,
            isFinishedStep: true,
            canJumpTo: stepIdReached >= 6
        }
    ];

    return (
        <React.Fragment>
            <Wizard
                isOpen
                onClose={ onClose }
                title="Create SCAP policy"
                description="Create a new policy for managing SCAP compliance"
                steps={ steps }
                onNext={ onNext }
            />
        </React.Fragment>
    );
};

CreatePolicy.propTypes = {
    benchmark: propTypes.string,
    osMajorVersion: propTypes.string,
    osMinorVersionCounts: propTypes.arrayOf(propTypes.shape({
        osMinorVersion: propTypes.number,
        count: propTypes.number
    })),
    complianceThreshold: propTypes.string,
    businessObjective: propTypes.object,
    dispatch: propTypes.func,
    isOpen: propTypes.bool,
    name: propTypes.string,
    onWizardFinish: propTypes.func,
    profile: propTypes.string,
    refId: propTypes.string,
    selectedRuleRefIds: propTypes.arrayOf(propTypes.string)
};

CreatePolicy.defaultProps = {
    isOpen: false
};

const selector = formValueSelector('policyForm');
export default connect(
    state => ({
        benchmark: selector(state, 'benchmark'),
        osMajorVersion: selector(state, 'osMajorVersion'),
        osMinorVersionCounts: selector(state, 'osMinorVersionCounts'),
        businessObjective: selector(state, 'businessObjective'),
        complianceThreshold: selector(state, 'complianceThreshold') || 100.0,
        name: selector(state, 'name'),
        profile: selector(state, 'profile'),
        refId: selector(state, 'refId'),
        selectedRuleRefIds: selector(state, 'selectedRuleRefIds')
    })
)(CreatePolicy);
