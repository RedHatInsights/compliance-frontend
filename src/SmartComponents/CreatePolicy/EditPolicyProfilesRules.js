import React, { useMemo, useLayoutEffect } from 'react';
import {
  propTypes as reduxFormPropTypes,
  formValueSelector,
  reduxForm,
} from 'redux-form';
import EmptyTable from '@redhat-cloud-services/frontend-components/EmptyTable';
import Spinner from '@redhat-cloud-services/frontend-components/Spinner';
import {
  Title,
  Text,
  TextContent,
  TextVariants,
  EmptyState,
  EmptyStateBody,
} from '@patternfly/react-core';
import { connect } from 'react-redux';
import { compose } from 'redux';
import propTypes from 'prop-types';
import { useQuery } from '@apollo/client';
import { StateViewWithError, StateViewPart } from 'PresentationalComponents';
import {
  TabbedRules,
  profilesWithRulesToSelection,
  tabsDataToOsMinorMap,
  extendProfilesByOsMinor,
} from 'PresentationalComponents/TabbedRules';
import * as Columns from '@/PresentationalComponents/RulesTable/Columns';
import useBenchmarksQuery from './hooks/useBenchmarksQuery';
import { PROFILES_QUERY } from './constants';

const getBenchmarkProfile = (benchmark, profileRefId) =>
  benchmark.profiles.find(
    (benchmarkProfile) => benchmarkProfile.refId === profileRefId
  );

const getBenchmarkBySupportedOsMinor = (benchmarks, osMinorVersion) =>
  benchmarks.find((benchmark) =>
    benchmark.latestSupportedOsMinorVersions?.includes(osMinorVersion)
  );

export const EditPolicyProfilesRules = ({
  policy,
  selectedRuleRefIds,
  change,
  osMajorVersion,
  osMinorVersionCounts,
  ruleValues,
}) => {
  const columns = [Columns.Name, Columns.Severity, Columns.Remediation];
  const osMinorVersions = osMinorVersionCounts
    .map((i) => i.osMinorVersion)
    .sort();
  const {
    data: benchmarksData,
    error: benchmarksError,
    loading: benchmarksLoading,
  } = useBenchmarksQuery({
    osMajorVersion,
    osMinorVersions,
  });

  const benchmarks = benchmarksData?.benchmarks?.nodes;

  const tabsData = useMemo(
    () =>
      osMinorVersionCounts
        .map(({ osMinorVersion, count: systemCount }) => {
          osMinorVersion = `${osMinorVersion}`;
          let profile;
          if (benchmarks) {
            const benchmark = getBenchmarkBySupportedOsMinor(
              benchmarks,
              osMinorVersion
            );
            if (benchmark) {
              profile = getBenchmarkProfile(benchmark, policy.refId);
              if (profile) {
                profile = {
                  ...profile,
                  benchmark: {
                    ...profile.benchmark,
                    ...benchmark,
                  },
                };
              }
            }
          }

          return {
            profile,
            systemCount,
            newOsMinorVersion: osMinorVersion,
          };
        })
        .filter(({ profile }) => !!profile),
    [osMinorVersionCounts, benchmarks, policy]
  );

  const profileToOsMinorMap = tabsDataToOsMinorMap(tabsData);
  const profileIds = Object.keys(profileToOsMinorMap);
  const filter = profileIds.map((i) => `id = ${i}`).join(' OR ');

  const {
    data: profilesData,
    error: profilesError,
    loading: profilesLoading,
  } = useQuery(PROFILES_QUERY, {
    variables: {
      filter,
    },
    skip: profileIds.length === 0,
    fetchPolicy: 'no-cache',
  });

  const error = benchmarksError || profilesError;
  const dataState = profileIds?.length > 0 ? profilesData : undefined;
  const loadingState = profilesLoading || benchmarksLoading ? true : undefined;
  const noRuleSets = !error && !loadingState && profileIds?.length === 0;
  const profiles = profilesData?.profiles.edges.map((p) => p.node);

  const setSelectedRuleRefIds = (newSelection) => {
    change('selectedRuleRefIds', newSelection);
  };

  const setRuleValues = (policyId, valueDefinition, valueValue) => {
    const newRuleValues = {
      ...(ruleValues || {}),
      [policyId]: {
        ...(ruleValues || {})[policyId],
        [valueDefinition.refId]: valueValue,
      },
    };

    change('ruleValues', newRuleValues);
  };

  useLayoutEffect(() => {
    if (!loadingState) {
      const profilesWithOs = extendProfilesByOsMinor(
        profiles,
        profileToOsMinorMap
      );
      const newSelection = profilesWithRulesToSelection(
        profilesWithOs,
        selectedRuleRefIds,
        { only: true }
      );
      setSelectedRuleRefIds(newSelection);
    }
  }, [JSON.stringify(profiles), loadingState]);

  return (
    <React.Fragment>
      <TextContent className="pf-u-pb-md">
        <Text component={TextVariants.h1}>Rules</Text>
        <Text>
          Customize your <b>{policy.name}</b> SCAP policy by including and
          excluding rules.
        </Text>
        <Text>
          Each release of RHEL is supported with a unique and specific version
          of the SCAP Security Guide (SSG). You must customize each version of
          SSG for each release of RHEL.
        </Text>
      </TextContent>

      <StateViewWithError
        stateValues={{
          error,
          data: dataState,
          loading: loadingState,
          noRuleSets,
        }}
      >
        <StateViewPart stateKey="noRuleSets">
          <EmptyState>
            <Title headingLevel="h1" size="xl">
              No rules can be configured
            </Title>
            <EmptyStateBody>
              The policy type selected does not exist for the systems and OS
              versions selected in the previous steps.
            </EmptyStateBody>
          </EmptyState>
        </StateViewPart>
        <StateViewPart stateKey="loading">
          <EmptyTable>
            <Spinner />
          </EmptyTable>
        </StateViewPart>
        <StateViewPart stateKey="data">
          <TabbedRules
            tabsData={tabsData}
            selectedRuleRefIds={selectedRuleRefIds}
            setRuleValues={setRuleValues}
            ruleValues={ruleValues}
            columns={columns}
            remediationsEnabled={false}
            selectedFilter
            level={1}
            setSelectedRuleRefIds={setSelectedRuleRefIds}
            ouiaId="RHELVersions"
            resetLink={true}
            rulesPageLink={true}
          />
        </StateViewPart>
      </StateViewWithError>
    </React.Fragment>
  );
};

EditPolicyProfilesRules.propTypes = {
  policy: propTypes.object,
  change: reduxFormPropTypes.change,
  osMajorVersion: propTypes.string,
  osMinorVersionCounts: propTypes.arrayOf(
    propTypes.shape({
      osMinorVersion: propTypes.number,
      count: propTypes.number,
    })
  ),
  selectedRuleRefIds: propTypes.array,
  ruleValues: propTypes.array,
};

const selector = formValueSelector('policyForm');

export default compose(
  connect((state) => ({
    policy: selector(state, 'profile'),
    osMajorVersion: selector(state, 'osMajorVersion'),
    osMinorVersionCounts: selector(state, 'osMinorVersionCounts'),
    selectedRuleRefIds: selector(state, 'selectedRuleRefIds'),
    ruleValues: selector(state, 'ruleValues'),
  })),
  reduxForm({
    form: 'policyForm',
    destroyOnUnmount: false,
    forceUnregisterOnUnmount: true,
  })
)(EditPolicyProfilesRules);
