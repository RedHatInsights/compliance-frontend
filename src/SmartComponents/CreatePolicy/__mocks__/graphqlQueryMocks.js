import { faker } from '@faker-js/faker';
import { graphqlResult } from '@/__factories__/helpers';
import { osMajorVersionsFromArray } from '@/__factories__/osMajorVersions.js';
import systems from '@/__factories__/systems.js';
import benchmarks from '@/__factories__/benchmarks.js';
import { countOsMinorVersions } from 'Store/Reducers/SystemStore';
import {
  SUPPORTED_PROFILES,
  BENCHMARKS_QUERY,
  BENCHMARKS_RULES_TREES_QUERY,
  BENCHMARKS_VALUE_DEFINITIONS_QUERY,
  PROFILES_QUERY,
} from '../constants';
import { GET_SYSTEMS_OSES } from '../../SystemsTable/constants';
import { BENCHMARK_QUERY } from '@/PresentationalComponents/TabbedRules/ProfileTabContent';

export default (osVersions) => {
  const osMajorVersions = osMajorVersionsFromArray(osVersions);
  const osMajorVersionToSelect = faker.helpers.arrayElement(osMajorVersions);

  const [profileToSelect, randomInUseProfile] = faker.helpers.arrayElements(
    osMajorVersionToSelect.profiles
  );
  const systemsToSelect = systems.buildList(
    1,
    {},
    { transient: { osVersions: profileToSelect.supportedOsVersions } }
  );

  const randomSystems = faker.helpers.arrayElements(systemsToSelect, {
    min: 1,
    max: systemsToSelect.length,
  });
  const randomSystemsCounts = countOsMinorVersions(randomSystems);

  const benchmarksForProfile = benchmarks.buildList(
    randomSystemsCounts.length,
    {},
    {
      transient: {
        profile: profileToSelect,
        osVersions: randomSystemsCounts.map(
          ({ osMinorVersion }) => `${osMinorVersion}`
        ),
      },
    }
  );

  const benchmarksFilter = `os_major_version = ${
    osMajorVersionToSelect.osMajorVersion
  } and latest_supported_os_minor_version ^ "${randomSystemsCounts
    .map(({ osMinorVersion }) => `${osMinorVersion}`)
    .sort()
    .join(',')}"`;

  const benchmarkResultOptions = {
    variables: {
      filter: benchmarksFilter,
    },
    inEdges: false,
  };

  const mocks = [
    graphqlResult(SUPPORTED_PROFILES, {
      osMajorVersions,
      profiles: randomInUseProfile ? [randomInUseProfile] : [],
    }),

    // This silence the apollo mock provider warning on the Systems step
    graphqlResult(
      GET_SYSTEMS_OSES,
      {},
      {
        variables: {
          filter: `os_major_version = ${
            osMajorVersionToSelect.osMajorVersion
          } AND os_minor_version ^ (${profileToSelect.supportedOsVersions
            .map((v) => v.split('.')[1])
            .join(',')})`,
        },
      }
    ),

    graphqlResult(
      BENCHMARKS_QUERY,
      {
        benchmarks: benchmarksForProfile,
      },
      benchmarkResultOptions
    ),
    graphqlResult(
      BENCHMARKS_RULES_TREES_QUERY,
      {
        benchmarks: benchmarksForProfile.map(({ id, ruleTree }) => ({
          id,
          ruleTree,
        })),
      },
      benchmarkResultOptions
    ),
    graphqlResult(
      BENCHMARKS_VALUE_DEFINITIONS_QUERY,
      {
        benchmarks: benchmarksForProfile,
      },
      benchmarkResultOptions
    ),
    graphqlResult(
      PROFILES_QUERY,
      {
        profiles: benchmarksForProfile,
      },
      {
        variables: {
          filter: `id = ${profileToSelect.id}`,
        },
      }
    ),

    graphqlResult(
      BENCHMARK_QUERY,
      {},
      {
        variables: {},
      }
    ),
  ];

  const mockResults = {
    mocks,
    osMajorVersionToSelect,
    profileToSelect,
    randomSystems,
    randomSystemsCounts,
  };

  return mockResults;
};
