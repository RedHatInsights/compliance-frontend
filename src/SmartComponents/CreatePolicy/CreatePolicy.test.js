import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { init } from 'Store';
import TestWrapper from '@/Utilities/TestWrapper';
import graphqlQueryMocks from './__mocks__/graphqlQueryMocks';

import CreatePolicy from './CreatePolicy';

const store = init().getStore();

const dispatchFieldChange = (store, field, payload) =>
  store.dispatch({
    type: '@@redux-form/CHANGE',
    meta: {
      form: 'policyForm',
      field,
    },
    payload,
  });

const checkStep = async (name, stepTests) => {
  expect(screen.getByRole('heading', { name })).toBeInTheDocument();

  stepTests && (await stepTests());
};

const nextStep = async (name, stepTests) => {
  await userEvent.click(
    screen.getByRole('button', { name: 'Next', enabled: true })
  );
  await checkStep(name, stepTests);
};

describe('CreatePolicy', () => {
  it('expect to render the Create Policy wizard', async () => {
    const {
      mocks,
      osMajorVersionToSelect,
      profileToSelect,
      randomSystems,
      randomSystemsCounts,
    } = graphqlQueryMocks(['7']);

    render(
      <TestWrapper mocks={mocks} store={store}>
        <CreatePolicy />
      </TestWrapper>
    );

    expect(screen.getByRole('button', { name: 'Next' })).toBeDisabled();

    await checkStep('Create SCAP policy', async () => {
      await screen.findByText('Operating system');
      await userEvent.click(
        screen.getByRole('option', {
          name: 'RHEL ' + osMajorVersionToSelect.osMajorVersion,
        })
      );

      expect(screen.getByText('Policy type')).toBeInTheDocument();

      await userEvent.type(
        // TODO Figure out how to set aria-label on filter input
        screen.getByRole('textbox', { name: 'text input' }),
        profileToSelect.name
      );

      await userEvent.click(
        screen.getByRole('radio', { name: 'Select row 0' })
      );
    });

    await nextStep('Details', async () => {
      expect(screen.getByText('Policy name')).toBeInTheDocument();
    });

    await nextStep('Systems', async () => {
      // Since we mock InventoryTable we need to dispatch a change manually
      dispatchFieldChange(store, 'systems', randomSystems);
      dispatchFieldChange(store, 'osMinorVersionCounts', randomSystemsCounts);
    });
    // TODO add proper rules data
    //     await nextStep('Rules', async () => {
    //       expect(screen.getByText(profileToSelect.name)).toBeInTheDocument();
    //
    //       await screen.findByText('SSG version');
    //
    //       profileToSelect.supportedOsVersions.forEach(async (osMinorVersion) => {
    //         expect(
    //           screen.getByRole('tabpanel', {
    //             name:
    //               'Rules for RHEL ' + osMajorVersionToSelect + '.' + osMinorVersion,
    //           })
    //         ).toBeInTheDocument();
    //       });
    //     });
  });
});
