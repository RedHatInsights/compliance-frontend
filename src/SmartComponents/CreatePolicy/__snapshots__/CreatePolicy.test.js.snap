// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePolicyForm expect to render the wizard 1`] = `
<Wizard
  appendTo={null}
  backButtonText="Back"
  cancelButtonText="Cancel"
  className="compliance"
  closeButtonAriaLabel="Close"
  description="Create a new policy for managing SCAP compliance"
  descriptionComponent="p"
  footer={null}
  hasDrawer={false}
  hasNoBodyPadding={false}
  height={null}
  hideClose={false}
  isDrawerExpanded={false}
  isNavExpandable={false}
  isOpen={true}
  mainAriaLabel={null}
  mainAriaLabelledBy={null}
  navAriaLabel={null}
  navAriaLabelledBy={null}
  nextButtonText="Next"
  onBack={[Function]}
  onClose={[Function]}
  onExpandDrawer={[Function]}
  onGoToStep={[Function]}
  onNext={[Function]}
  startAtStep={1}
  steps={
    Array [
      Object {
        "component": <Memo(Connect(ReduxForm)) />,
        "enableNext": false,
        "id": 1,
        "name": "Create SCAP policy",
      },
      Object {
        "canJumpTo": false,
        "component": <Memo(Connect(ReduxForm)) />,
        "enableNext": false,
        "id": 2,
        "name": "Details",
      },
      Object {
        "canJumpTo": false,
        "component": <Memo(Connect(ReduxForm)) />,
        "enableNext": false,
        "id": 3,
        "name": "Systems",
      },
      Object {
        "canJumpTo": false,
        "component": <Memo(Connect(ReduxForm)) />,
        "enableNext": false,
        "id": 4,
        "name": "Rules",
      },
      Object {
        "canJumpTo": false,
        "component": <Memo(Connect(ReviewCreatedPolicy)) />,
        "id": 5,
        "name": "Review",
        "nextButtonText": "Finish",
      },
      Object {
        "canJumpTo": false,
        "component": <Memo(Connect(ReduxForm))
          onWizardFinish={[Function]}
        />,
        "id": 6,
        "isFinishedStep": true,
        "name": "Finished",
      },
    ]
  }
  title="Create SCAP policy"
  width={1220}
/>
`;

exports[`CreatePolicyForm expect to render the wizard with enableNext on systems 1`] = `true`;
